dataset.py

```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  )
+
+  return PartitionedDataset(
+    train=train_split,
+    test=test_split,
+    validation=validation_split
+  )
+
+def _polygon(
+  coordinates: list[list]) -> Polygon:
+  '''
+  Given an unordered list of longitude and latitude coordinates
+  represented by tuples, returns a polygon with the shape of the
+  coordinates.
+  '''
+  if len(coordinates) < 3:
+    return None
+  # Compute centroid
+  centroid = (sum([c[0] for c in coordinates])/len(coordinates),
+              sum([c[1] for c in coordinates])/len(coordinates))
+  # Sort by polar angle. This arranges the coordinates so that they
+  # delimit an outward shape.
+  sorted_coordinates = list(coordinates)
+  sorted_coordinates.sort(
+    key=lambda c: math.atan2(
+      c[1] - centroid[1],
+      c[0] - centroid[0])
+  )
+  return Polygon(sorted_coordinates)
+
+def _valid_polygons(
+  train_polygon: Polygon,
+  validation_polygon: Polygon,
+  test_polygon: Polygon) -> bool:
+  '''
+  Returns true if the train, validation, and test polygons
+  have a larger area than or equal to 0.1 and they don't overlap
+  over each other. Returns false otherwise. 
+  '''
+  small_polygon = False
+  for polygon in [train_polygon, validation_polygon, test_polygon]:
+    if polygon.area < 0.1:
+      return False
+
+  if (train_polygon.intersects(validation_polygon) or
+     train_polygon.intersects(test_polygon) or
+     validation_polygon.intersects(test_polygon)):
+    return False
+
+  return True
+
+def _shuffled_unique_coordinates(
+  sample_data: pd.DataFrame,
+  strategy: FurthestPointsPartitionStrategy) -> list[list[float]]:
+  '''
+  Returns shuffled unique coordinates from sample_data that match
+  the expected train, validation and test fractions.
+  '''
+  unique_coordinates_df = sample_data.groupby(
+    by=['long', 'lat']
+  ).first().reset_index()
+  shuffled_coordinates_df = unique_coordinates_df.sample(frac=1.0,
+    random_state=strategy.random_seed)
+  coordinates = list(
+    shuffled_coordinates_df[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+  assert(len(coordinates) >= strategy.top_n)
+  min_size = (int(len(coordinates) * strategy.train_fraction) +
+    int(len(coordinates) * strategy.validation_fraction) +
+    int(len(coordinates) * strategy.test_fraction))
+  assert(int(len(coordinates) * strategy.train_fraction) > 0)
+  assert(int(len(coordinates) * strategy.validation_fraction) > 0)
+  assert(int(len(coordinates) * strategy.test_fraction) > 0)
+  assert(
+    min_size <= len(coordinates)
+  ), (f"You need {min_size} aka {int(len(coordinates) * strategy.train_fraction)} train + "
+    f"{int(len(coordinates) * strategy.validation_fraction)} validation + "
+    f"{int(len(coordinates) * strategy.test_fraction)} test samples but you have a sample of {len(coordinates)}")
+  
+  return coordinates
+
+def _maybe_partition_furthest_points(
+  sample_data: pd.DataFrame,
+  furthest_coordinates: list[list[float]],
```

Comment:
Can you add an Args: block in the comments that explain what the list[list[float]] is for?

I think maybe this is supposed to be a `float **furthest_coordinates` which would make a lot of sense in C but in Python we would prefer a `List[Tuple[float, float]]`.

Side-note: Does your C++ team do this? Because in C++ I would usually do a `std::vector<std::pair<float, float>>` or `std::vector<std::tuple<float, float>>` or maybe even create a type and do `std::vector<CoordinatePair>` if the meaning of the former was unclear.

Reply:
Created Coordinate class.